pkg/bpf/perf.go

```diff
@@ -56,57 +68,97 @@ struct event_sample {
 };
 
 struct read_state {
-	size_t raw_size;
-	void *base, *begin, *end, *head;
+	void *base;
+	uint64_t raw_size;
+	uint64_t last_size;
 };
 
-int perf_event_read_init(int page_count, int page_size, void *_header, void *_state)
+// Contract with kernel/user space on perf ring buffer:
+//
+//   kernel                             user
+//
+//   if (LOAD ->data_tail) {            LOAD ->data_head
+//                      (A)             smp_rmb()       (C)
+//      STORE $data                     LOAD $data
+//      smp_wmb()       (B)             smp_mb()        (D)
+//      STORE ->data_head               STORE ->data_tail
+//   }
+
+static inline uint64_t perf_read_head(struct perf_event_mmap_page *up)
+{
+	uint64_t data_head = READ_ONCE_64(up->data_head);
+	rmb();
+	return data_head;
+}
+
+static inline void perf_write_tail(struct perf_event_mmap_page *up,
+				   uint64_t data_tail)
+{
+	mb();
+	up->data_tail = data_tail;
+}
+
+int perf_event_read_init(int page_count, int page_size, void *_page,
+			 void *_state)
 {
-	volatile struct perf_event_mmap_page *header = _header;
+	struct perf_event_mmap_page *up = _page;
 	struct read_state *state = _state;
-	uint64_t data_tail = header->data_tail;
-	uint64_t data_head = *((volatile uint64_t *) &header->data_head);
+	uint64_t data_tail = up->data_tail;
 
-	__sync_synchronize();
-	if (data_head == data_tail)
+	if (perf_read_head(up) == data_tail)
 		return 0;
 
-	state->head = (void *) data_head;
 	state->raw_size = page_count * page_size;
-	state->base  = ((uint8_t *)header) + page_size;
-	state->begin = state->base + data_tail % state->raw_size;
-	state->end   = state->base + data_head % state->raw_size;
+	state->base = ((uint8_t *)up) + page_size;
+	state->last_size = 0;
 
-	return state->begin != state->end;
+	return 1;
 }
 
-int perf_event_read(void *_state, void *buf, void *_msg, void *_sample, void *_lost)
+int perf_event_read(void *_page, void *_state, void *_buf, void *_msg,
+		    void *_sample, void *_lost, int page_size)
 {
-	void **msg = (void **) _msg;
+	struct perf_event_mmap_page *up = _page;
+	struct read_state *state = _state;
 	void **sample = (void **) _sample;
 	void **lost = (void **) _lost;
-	struct read_state *state = _state;
-	struct event_sample *e = state->begin;
-
-	if (state->begin == state->end)
+	void **msg = (void **) _msg;
+	void **buf = (void **) _buf;
+	uint64_t e_size, data_tail;
+	struct event_sample *e;
+	void *begin;
+
+	data_tail = up->data_tail + state->last_size;
```

Comment:
Looking at this, I was thinking that there needs to be a guarantee that `(UINT64_MAX % page_size) == 0`, otherwise the modulo operations below would mess up when the uint64 head/tail overflows. Is that right? How is that enforced?

Reply:
It's enforced here in `ring_buffer_has_space()` https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/events/ring_buffer.c#n160 , I'll double check tomorrow, but my understanding is that it should be fine given the diff is relative (not diverting from perf/others either).